File: 2022/day02.py
def score(p1, p2):
    if (p1 - 1) % 3 == p2:
        return p2 + 1
    elif (p2 - 1) % 3 == p1:
        return p2 + 7
    return p2 + 4


with open("input_files/day02", "r") as f:
    data = f.read().splitlines()
    data = [line.split() for line in data]
    data = [(ord(p1) - ord("A"), ord(p2) - ord("X")) for p1, p2 in data]

moves = [(p1, (p1 + p2 - 1) % 3) for p1, p2 in data]

print(sum(score(p1, p2) for p1, p2 in data))
print(sum(score(p1, p2) for p1, p2 in moves))

File: koltin/src/main/kotlin/y2022/day02.kt
package y2022

import readLines

fun main() {
    val data = readLines("y2022/day02")
        .map { it.toCharArray() }
        .map { (p1, _, p2) -> p1 - 'A' to p2 - 'X' }

    val moves = data.map { (p1, p2) -> p1 to (p1 + p2 - 1).mod(3) }

    println(data.sumOf { it.score() })
    println(moves.sumOf { it.score() })
}

private fun Pair<Int, Int>.score(): Int {
    val (p1, p2) = this
    return if ((p1 - 1).mod(3) == p2) {
        p2 + 1
    } else if ((p2 - 1).mod(3) == p1) {
        p2 + 7
    } else {
        p2 + 4
    }
}
